### DISCUSSION TITLE
about how to use start_background_task?

### DISCUSSION THREAD

**Discussion by ufoozhenghao** [miguelgrinberg/Flask-SocketIO#2023] [Q&A] [answered]
Functional Requirement: When the client successfully connects, Python should send a random number to the HTML at a frequency of once per second. 
However, I found that if I add a 'while true' loop after using the start_background_task function, the program will not run successfully. How can I solve this?

```
#app.py
import random
import time

from flask import Flask, render_template
from flask_socketio import SocketIO, emit

app = Flask(__name__)
# socketio = SocketIO(app, async_mode='eventlet')
socketio = SocketIO(app, cors_allowed_origins="*")


@app.route('/')
def index():
    return "Hello, World!"


@socketio.on('connect')
def handle_connect():
    print('Client connected')
    socketio.start_background_task(background_task)


@socketio.on('disconnect')
def handle_disconnect():
    print('Client disconnected')


def background_task():
    while True:
        number = random.randint(1, 100)
        socketio.emit('message', {'data': number})
        time.sleep(1)  # Pause for a second before next number


if __name__ == '__main__':
    socketio.run(app, port=5010)

```

my js code:
```
// io
const socket = io.connect('http://localhost:5010');
console.log(socket)
socket.on('message', (data)=>{
  console.log(data)
});
socket.on('connect', function(msg) {
  console.log('Connected to the server');
  console.log(msg);
});
socket.on('connect_error', function(error) {
  console.log('Connection failed: ', error);
});
```

**Comment by miguelgrinberg** [comment]
> the program will not run successfully

What does this mean exactly?

**Reply by ufoozhenghao** [reply]
i can't get data in client side which i use html

**Reply by miguelgrinberg** [reply]
I'm sorry but I still don't understand what problem you are reporting. Please provide logs, error messages, stack traces, etc.

**Reply by ufoozhenghao** [reply]
thank you for your reply.

Requirement:
when the client side connected or emit to server(which use flask_socketio), the server will continue sending data to the client. I need to ensure the real-time nature of the data.

I tried to use `while true`loop in `socketio.start_background_task`, but it need add`socketio.sleep(5)`to work. If i dont use `socketio.sleep(5)`, the code dons't work(won't emit message to client). So,Is there any method to deliver server data to the client in real-time? I don't want use `socketio.sleep(5)`

here is my code:
```
from flask import Flask, render_template
from flask_socketio import SocketIO,emit
from threading import Lock
import random
async_mode = None
app = Flask(__name__)
app.config['SECRET_KEY'] = 'secret!'
socketio = SocketIO(app)
thread = None
thread_lock = Lock()

@app.route('/')
def index():
    return render_template('index.html')

@socketio.on('connect', namespace='/test_conn')
def test_connect():
    global thread
    with thread_lock:
        if thread is None:
            thread = socketio.start_background_task(target=background_thread)


# here i use while loop and socketio.sleep(5)
def background_thread():
    while True:
        socketio.sleep(5)
        t = random.randint(1, 100)
        socketio.emit('server_response',
                      {'data': t},namespace='/test_conn')

if __name__ == '__main__':
    socketio.run(app, debug=True)
```

**Reply by miguelgrinberg** [reply, marked as answer]
Two options:

- Use `socket.sleep(0)`, there is no requirement to pause for 5 seconds. A sleep with at least 0 is required when using greenlets (eventlet or gevent) to maintain concurrency.
- If you don't want to sleep at all, use the `threading` async_mode, which implements concurrency with standard threads.
